### PR TITLE
Test input variable distributions against UKMOD

### PR DESCRIPTION
**Pull request by nikhilwoodruff**
This PR will test that the variable inputs to openfisca-uk have similar distributions to their equivalents in UKMOD. I expect most of these will not be very hard tests to pass, but I think it's a good idea to set this testing structure for the rest of the microdata testing (`openfisca-uk-data` and `openfisca-uk`), so if we find any discrepancies down the line we can narrow it down quickly.

Tests for each variable:
* Quantiles 10, 20, ... 90
* Aggregates
* Nonzero counts
* Average error (openfisca-uk household <-> UKMOD input household) in nonzero households
* Nonzero agreement rate

### CODE REVIEW

**Comment by nikhilwoodruff** [comment]
@MaxGhenis all the tests now pass (Pandas-ified and adjusted the implementations slightly to more closely match the UKMOD data requirements document - out of the >550 tests, <5 needed slightly increased bounds to pass). I'll add these add issues but they mostly seem justifiable: e.g. the Winter Fuel Allowance data in UKMOD has a few households receiving £400 per year, and I couldn't figure out why because it's [at the household leve](https://www.microsimulation.ac.uk/wp-content/uploads/2020/10/cempa7-20.pdf#page=11) and the benefits table in the raw FRS has a maximum payment of £300.

File: openfisca_uk_data/datasets/frs/frs.py [hunk 1]
```diff
@@ -1,87 +1,635 @@
...
+    CT_imputed = hh_CT_mean
+    council_tax = pd.Series(
+        np.where(
+            household.CTANNUAL.isna(), max_(CT_imputed, 0), household.CTANNUAL
+        )
+    )
+    frs["council_tax"] = council_tax
+
+    frs["council_tax_band"] = (
+        household.CTBAND.fillna(4)
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:288]: Add comment - is `D` the most common?

File: openfisca_uk_data/datasets/frs/frs.py [hunk 2]
```diff
@@ -1,87 +1,635 @@
...
+        pension (DataFrame)
+        job (DataFrame)
+        account (DataFrame)
+        household (DataFrame)
+        oddjob (DataFrame)
+        year (DataFrame)
+    """
+    frs["employment_income"] = person.INEARNS * 52
+
+    frs["pension_income"] = (
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:327]: split this into multiple statements

File: openfisca_uk_data/datasets/frs/frs.py [hunk 3]
```diff
@@ -1,87 +1,635 @@
...
+        (pension.PENPAY * (pension.PENPAY > 0)).groupby(pension.person_id).sum().reindex(index=person.index).fillna(0)
+        + (pension.PTAMT * ((pension.PTINC == 2) & (pension.PTAMT > 0))).groupby(pension.person_id).sum().reindex(index=person.index).fillna(0)
+        + (pension.POAMT * (((pension.POINC == 2) | (pension.PENOTH == 1)) & (pension.POAMT > 0))).groupby(pension.person_id).sum().reindex(index=person.index).fillna(0)
+    ) * 52
+
+    # Add self-employed income (correcting one person in 2018)
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:333]: could you document this in an issue?

File: openfisca_uk_data/datasets/frs/frs.py [hunk 4]
```diff
@@ -1,87 +1,635 @@
...
+    frs["tax_free_savings_income"] = (
+        account.ACCINT[account.ACCOUNT == 21]
+        .groupby(account.person_id)
+        .sum()
+        .reindex(index=person.index)
+        .fillna(0)
+        * 52
+    )
+    frs["savings_interest_income"] = (
+        (account.ACCINT * np.where(account.ACCTAX == 1, 1.25, 1))[
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:357]: could you make `1.25` a constant and document it

File: openfisca_uk_data/datasets/frs/frs.py [hunk 5]
```diff
@@ -1,87 +1,635 @@
...
+            account.ACCOUNT.isin((1, 3, 5, 27, 28))
+        ]
+    ).groupby(account.person_id).sum().reindex(index=person.index).fillna(
+        0
+    ) * 52 + frs[
+        "tax_free_savings_income"
+    ][
+        ...
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:365]: ?

File: openfisca_uk_data/datasets/frs/frs.py [hunk 6]
```diff
@@ -1,87 +1,635 @@
...
+    ]
+    frs["dividend_income"] = (
+        account.ACCINT * np.where(account.INVTAX == 1, 1.25, 1)
+    )[
+        ((account.ACCOUNT == 6) & (account.INVTAX == 1))
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:370]: add comment, and/or split into multiple statements

File: openfisca_uk_data/datasets/frs/frs.py [hunk 7]
```diff
@@ -1,87 +1,635 @@
...
+                    person.household_id
+                ].values,
+                index=person.index,
+            ).fillna(0)
+        )
+        + person.CVPAY
+        + person.ROYYR1
+    ) * 52
+
+    # Discrepancy in maintenance income (UKMOD appears to use last amount rather than usual, with "not usual" answer):
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:394]: please document in an issue

File: openfisca_uk_data/datasets/frs/frs.py [hunk 8]
```diff
@@ -1,87 +1,635 @@
...
+
+    # Add self-employed income (correcting one person in 2018)
+    seincamt = (
+        (job.SEINCAMT.groupby(job.person_id).sum())
+        .reindex(person.index)
+        .fillna(0)
+    )
+    frs["self_employment_income"] = (
+        np.where(
+            (year == 2018) & (person.index == 806911),
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:341]: make 806911 a constant like the maintenance adjustment records below

File: openfisca_uk_data/datasets/frs/frs.py [hunk 9]
```diff
@@ -1,87 +1,635 @@
...
+    ).sum().reindex(
+        index=person.index
+    ).fillna(
+        0
+    ) * 52
+    is_head = person.HRPID == 1
+    frs["property_income"] = (
+        is_head
+        * (
+            pd.Series(
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:383]: separate this into a new expression

File: openfisca_uk_data/datasets/frs/frs.py [hunk 10]
```diff
@@ -1,87 +1,635 @@
...
+        1033500,
+        322100,
+        42100,
+        590400,
+        645300,
+        1901400,
+        388000,
+    )
+
+    frs["maintenance_income"] = (
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:407]: split this up

File: openfisca_uk_data/datasets/frs/frs.py [hunk 11]
```diff
@@ -1,87 +1,635 @@
...
+                    person.MNTUSAM2,
+                    person.MNTAMT2,
+                )
+            ).fillna(0),
+            0,
+        )
+    ) * 52
+
+    frs["miscellaneous_income"] = (
+        oddjob.OJAMT[oddjob.OJNOW == 1]
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:435]: split this out

File: openfisca_uk_data/datasets/frs/frs.py [hunk 12]
```diff
@@ -1,87 +1,635 @@
...
+        .groupby(oddjob.person_id)
+        .sum()
+        .reindex(person.index)
+        .fillna(0)
+        + person.ALLPAY2
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:440]:
```suggestion
        + person[["ALLPAY2", "ROYYR2", "ROYYR3", "CHAMTERN", "CHAMTTST"]].sum(axis=1)
```
could also name those columns as a constant first

File: openfisca_uk_data/datasets/frs/frs.py [hunk 13]
```diff
@@ -1,87 +1,635 @@
...
+        frs (h5py.File)
+        person (DataFrame)
+        household (DataFrame)
+        maintenance (DataFrame)
+        mortgage (DataFrame)
+        childcare (DataFrame)
+        pen_prov (DataFrame)
+    """
+    frs["maintenance_expenses"] = (
+        pd.Series(
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:557]: i think you could save some code by making this pattern a function, e.g.
```
def sum_to_level(vals, dataset, level):
    reindex = person.index if level == "person" else household.index
    return pd.Series(vals).groupby(dataset[level + "_id"].sum().reindex(reindex).fillna(0)
```
i.e. this then becomes
```
frs["maintenance_expenses"] = sum_to_level(
    np.where(maintenance.MRUS == 2, maintenance.MRUAMT, maintenance.MRAMT),
    maintenance,
    "person") * 52

File: openfisca_uk_data/datasets/frs/frs.py [hunk 14]
```diff
@@ -1,87 +1,635 @@
...
+        .fillna(0)
+    ) * 52
+
+    frs["private_pension_contributions"] = (
+        pen_prov.PENAMT[pen_prov.STEMPPEN.isin((5, 6))]
+        .groupby(pen_prov.person_id)
+        .sum()
+        .reindex(person.index)
+        .fillna(0)
+        .clip(0, pen_prov.PENAMT.quantile(0.95))
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs.py:603]: is this what ukmod does?

Comment (nikhilwoodruff) [reply openfisca_uk_data/datasets/frs/frs.py:603]: I think so, from the input microdata documentation:
```
gen xpp=penamt*52/12 if stemppen==5|stemppen==6 
 su xpp if xpp>0, d
 replace xpp = r(p95) if xpp > r(p95)
```

Comment (MaxGhenis) [reply openfisca_uk_data/datasets/frs/frs.py:603]: Do they say why they clip that variable specifically? It'd be good to document this choice somewhere.

**Comment by nikhilwoodruff** [comment]
@MaxGhenis think I got all the comments (need to file issues for the discrepancy adjustments).

**Comment by nikhilwoodruff** [comment]
Requires https://github.com/PolicyEngine/openfisca-uk/pull/213

File: Makefile [hunk 15]
```diff
@@ -2,10 +2,12 @@ all:
 	pip install wheel
 	python setup.py sdist bdist_wheel
 install:
-	pip install git+https://github.com/PSLmodels/openfisca-uk
+	pip install git+https://github.com/nikhilwoodruff/openfisca-uk
```
Comment (MaxGhenis) [inline Makefile:5]:
```suggestion
	pip install git+https://github.com/PolicyEngine/openfisca-uk
```

File: openfisca_uk_data/datasets/frs/frs_was_imputation.py [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+import pandas as pd
+import microdf as mdf
+import sklearn
+import synthimpute as si
+import h5py
+import numpy as np
+
+
+@dataset
+class FRS_WAS_Imputation:
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs_was_imputation.py:12]: Should we call this something else? I think we'd ideally make a master FRS-based dataset with WAS imputations, SPI adjustments, LCF imputations, etc.

File: openfisca_uk_data/datasets/frs/frs_was_imputation.py [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+        "self_employment_income",
+        "investment_income",
+        "num_bedrooms",
+        "council_tax",
+        "is_renting",
+        # "region"  # TODO: add region mapping from WAS int.
+    ]
+
+    IMPUTE_COLS = [
+        "est_land",  # Total wealth.
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs_was_imputation.py:59]:
```suggestion
        "est_land",  # Estimated land value based on property and corporate wealth.
```

File: openfisca_uk_data/datasets/frs/frs_was_imputation.py [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+        "num_adults",
+        "num_children",
+        "pension_income",
+        "employment_income",
+        "self_employment_income",
+        "investment_income",
+        "num_bedrooms",
+        "council_tax",
+        "is_renting",
+        # "region"  # TODO: add region mapping from WAS int.
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs_was_imputation.py:55]: move to an issue

File: openfisca_uk_data/datasets/frs/frs_was_imputation.py [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+    ]
+
+    IMPUTE_COLS = [
+        "est_land",  # Total wealth.
+    ]
+
+    # TODO: Test imputation quantiles on a hold out sample.
```
Comment (MaxGhenis) [inline openfisca_uk_data/datasets/frs/frs_was_imputation.py:62]: we can remove this and report model quality on a holdout set somewhere else in the future

File: openfisca_uk_data/datasets/frs/frs_was_imputation.py (hunk 16 above)
Comment (nikhilwoodruff) [reply openfisca_uk_data/datasets/frs/frs_was_imputation.py:12]: In future yes, OK to merge for now though?

File: openfisca_uk_data/datasets/frs/frs_was_imputation.py (hunk 18 above)
Comment (nikhilwoodruff) [reply openfisca_uk_data/datasets/frs/frs_was_imputation.py:55]: Moved

File: openfisca_uk_data/datasets/frs/frs_was_imputation.py (hunk 19 above)
Comment (nikhilwoodruff) [reply openfisca_uk_data/datasets/frs/frs_was_imputation.py:62]: OK yep, leaving in until then

**Pull request merged**